File: ruby/allergies/allergies.rb
# frozen_string_literal: true

class Allergies
  attr_reader :list

  ITEMS = %w[eggs peanuts shellfish strawberries tomatoes chocolate pollen cats].freeze

  def initialize(score)
    @list = allergy_list(score)
  end

  def allergic_to?(item)
    list.include?(item)
  end

  private

  def allergy_list(score)
    reverse_indexes = (0..ITEMS.size - 1).to_a.reverse
    reverse_indexes.each_with_object(list: [], score: score % 256) do |i, acc|
      if acc[:score] >= 2**i
        acc[:list] << ITEMS[i]
        acc[:score] -= 2**i
      end
    end[:list]
  end
end

File: python/allergies/allergies.py
class Allergies:

    def __init__(self, score):
        self.score = score

    def allergic_to(self, item):
        return item in self.lst

    @property
    def lst(self):
        allergens = ['eggs', 'peanuts', 'shellfish', 'strawberries', 'tomatoes', 'chocolate', 'pollen', 'cats']
        score = self.score % 256
        result = []
        for power in list(reversed(range(len(allergens)))):
            if score >= 2**power:
                result.append(allergens[power])
                score -= 2**power

        return result
